cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDynamicDirectedGraph.scala

```diff
@@ -0,0 +1,206 @@
+package com.twitter.cassovary.graph
+
+import com.twitter.cassovary.graph.node.DynamicNode
+import it.unimi.dsi.fastutil.ints.IntArrayList
+import scala.collection.mutable.ArrayBuffer
+import StoredGraphDir._
+import com.twitter.cassovary.util.FastUtilUtils
+
+/**
+ * A dynamic directed graph, implemented using an ArrayBuffer of IntArrayList
+ * (from the fastutil library).  If n nodes are used, O(n) objects are created,
+ * independent of the number of edges.
+ *
+ * For efficiency, it's recommended that ids be sequentially numbered, as
+ * an internal array is stored which is longer than the maximum id seen.
+ *
+ * This class is not threadsafe.  If multiple thread access it simultaneously
+ * and at least one mutates it, synchronization should be used.
+ */
+class ArrayBasedDynamicDirectedGraph(override val storedGraphDir: StoredGraphDir)
+    extends DynamicDirectedGraph {
+  // outboundLists(id) contains the outbound neighbors of the given id,
+  // or null if the id is not in this graph.
+  // If we aren't storing outbound neighbors, outboundLists will always remain size 0.
+  private val outboundLists = new ArrayBuffer[IntArrayList]
+  // (See above note on outboundLists)
+  private val inboundLists = new ArrayBuffer[IntArrayList]
+
+  private var nodeCount_ = 0
+
+  def this(dataIterator: Iterator[NodeIdEdgesMaxId],
+            storedGraphDir: StoredGraphDir) {
+    this(storedGraphDir)
+    for (nodeData <- dataIterator) {
+      val id = nodeData.id
+      getOrCreateNode(id)
+      nodeData.edges map getOrCreateNode
+      storedGraphDir match {
+        case OnlyOut => outboundLists(id).addAll(IntArrayList.wrap(nodeData.edges))
+        case OnlyIn => inboundLists(id).addAll(IntArrayList.wrap(nodeData.edges))
+        case BothInOut => nodeData.edges map { addEdgeAllowingDuplicates(id, _) } // Duplicates shouldn't exist, but allow them for efficiency.
+      }
+    }
+  }
+
+  def this(iteratorSeq: Seq[() => Iterator[NodeIdEdgesMaxId]],
+           storedGraphDir: StoredGraphDir) {
+    this((iteratorSeq map { _()} ).iterator.flatten, storedGraphDir)
```

Comment:
Good idea to just test it.  If I execute the following script:
```scala
def makeIterator(name: String): Iterator[Int] = new Iterator[Int] {
  var current = 0
  def next(): Int =  {
    println("iterator " + name + " returning " + current)
    current += 1
    current - 1
  }
  def hasNext(): Boolean = (current < 5)
}
val iteratorSeq: Seq[() => Iterator[Int]] = Seq(
    () => makeIterator("one"),
    () => makeIterator("two"),
    () => makeIterator("three"))

val flatMapResult = iteratorSeq flatMap { _()}
```
Then it prints 15 lines of the form "iterator _ returning _"
Wheras if I replace the last line with
`val moreAwkwardResult = (iteratorSeq map { _()} ).iterator.flatten`
then it prints nothing.